Enable write access to models initialized from arrays in the interpreter

tronical opened the pull request:
A model property such as

    property <[Foo]> model: [ ... ];

would create a VecModel in Rust and C++, which in turn can be written to
with an array index expression in .60. If the same .60 code was running
in the interpreter, a Value::Array would be created instead. We don't
support writing into Value::Array by index because we don't have
tracking in place for that, to update dependent bindings.

To fix this inconsistency, this patch changes the interpreter to
allocate a mutable VecModel as well.

ogoffart reviewed (commented):
We should deprecate (or remove?, since next release is probably breaking) the Value::Array

tronical commented:
> We should deprecate (or remove?, since next release is probably breaking) the Value::Array

I've already edited the corresponding ticket and added a note about deprecating/removing `Value::Array`.

That said, I see a bigger problem in the failing test: We can't represent a model created in the interpreter in C++ (and don't have `Value::to_model()`). We could do the same as in JS: Create a SharedVector/Array out of the model on-the-fly though. What do you think?

tronical commented on api/sixtyfps-node/native/lib.rs:
```diff
@@ -276,6 +276,15 @@ fn to_js_value<'cx>(
             &format!("#{:02x}{:02x}{:02x}{:02x}", c.red(), c.green(), c.blue(), c.alpha()),
         )
         .as_value(cx),
+        Value::Model(model) => {
```
That's a good point, I agree. Will look into that.

I think `JsModel` is used & created when assigning a "model like" object from JS to a model property in .60. So in a sense most of our tests (like `write_to_model.60`) that create a `sixtyfps::ArrayModel` in JS and then assign it to a model property cover that.

> There is also an occurence in api/sixtyfps-node/native/lib.rs:180 that needs to be changed to use a Model

Indeed. I wonder why this didn't break anything. I guess because we continue to support that within the interpreter, i.e. `ValueModel` in the interpreter supports `Value::Array`.

tronical replied:
> We should try to do a as_dyn of a JsModel and then return the inner object.

I've made a commit that tests and fixes the round-tripping.

tronical commented:
> > We should deprecate (or remove?, since next release is probably breaking) the Value::Array
> 
> I've already edited the corresponding ticket and added a note about deprecating/removing `Value::Array`.
> 
> That said, I see a bigger problem in the failing test: We can't represent a model created in the interpreter in C++ (and don't have `Value::to_model()`). We could do the same as in JS: Create a SharedVector/Array out of the model on-the-fly though. What do you think?

I've fixed this now by creating a `SharedVectorModel<Value>` instead of a `VecModel` in the interpreter. That should come at no additional "cost", but it allows a cheap extraction of the values to a `Value::Array` in C++, which is the sensible representation for models created from arrays in `.60`.

I need to rebase & squash this, but meanwhile I'll look first into the js_model feedback.

tronical commented:
This is becoming messy in terms of the conflicts, so I'll go ahead and rebase this.

ogoffart commented on api/sixtyfps-node/native/lib.rs:
```diff
@@ -258,15 +262,30 @@ fn to_js_value<'cx>(
+            if let Some(js_model) = model.as_any().downcast_ref::<js_model::JsModel>() {
+                js_model.get_object(cx, persistent_context)?.as_value(cx)
+            } else {
+                // TODO: this should probably create a proxy object instead of extracting the entire model. On the other hand
+                // we should encounter this only if the model was created in .60, which is when it'll be an array
+                // of values.
```
But this means that the model from .60 are different.
If you have a model set from .js, you can do
```js
instance.the_model[idx] = something;
```
and that will update the UI.

But if the model is comming from .60 and so this code path is run, then one will be operating on a copy of the model instead.
So we are back to the inconsistancy.

tronical replied:
Yes, that's a good point. Previously this code caused a panic, now it will silently "fail" (not update the UI).

ogoffart reviewed (approved):
So first of all, we can remove the Value::Array (i don't think it is ever created anymore), and ValueType::Array.

Then there is the problem that we don't return proxy model, in JS and C++
I guess this is a bit of work to implement and so we can just pospone that.

But otherwise, i think this can already go in.

tronical commented:
> So first of all, we can remove the Value::Array (i don't think it is ever created anymore), and ValueType::Array.

Indeed, we don't need it in Rust anymore - it's only needed in C++ in order to get access to `property <[Foo]> model: [ ... ];` from C++, i.e. we'd convert that to a `Value::Type::Array` and `Value::to_array()` will return a `SharedVector<Value>`.

> Then there is the problem that we don't return proxy model, in JS and C++ I guess this is a bit of work to implement and so we can just pospone that.

Indeed.

> But otherwise, i think this can already go in.

Okay. I'll create follow-up tickets for the remaining `Value::Array` issues (C++ and JS).
